crates/matrix-sdk-ui/src/timeline/inner.rs

```diff
@@ -727,117 +728,152 @@ impl<P: RoomDataProvider> TimelineInner<P> {
-                };
+        if retry_indices.is_empty() {
+            return;
+        }
 
-                tracing::Span::current().record("session_id", session_id);
+        debug!("Retrying decryption");
+
+        let track_read_receipts = self.track_read_receipts;
+        let room_data_provider = self.room_data_provider.clone();
+        let push_rules_context = room_data_provider.push_rules_and_context().await;
+
+        let do_retry = async move {
+            let retry_one = |item: Arc<TimelineItem>| {
+                let decryptor = decryptor.clone();
+                let should_retry = &should_retry;
+                async move {
+                    let event_item = item.as_event()?;
+
+                    let session_id = match event_item.content().as_unable_to_decrypt()? {
+                        EncryptedMessage::MegolmV1AesSha2 { session_id, .. }
+                            if should_retry(session_id) =>
+                        {
+                            session_id
+                        }
+                        EncryptedMessage::MegolmV1AesSha2 { .. }
+                        | EncryptedMessage::OlmV1Curve25519AesSha2 { .. }
+                        | EncryptedMessage::Unknown => return None,
+                    };
+
+                    tracing::Span::current().record("session_id", session_id);
+
+                    let Some(remote_event) = event_item.as_remote() else {
+                        error!("Key for unable-to-decrypt timeline item is not an event ID");
+                        return None;
+                    };
+
+                    tracing::Span::current().record("event_id", debug(&remote_event.event_id));
+
+                    match decryptor.decrypt_event_impl(&remote_event.original_json).await {
+                        Ok(event) => {
+                            trace!(
+                                "Successfully decrypted event that previously failed to decrypt"
+                            );
+                            Some(event)
+                        }
+                        Err(e) => {
+                            info!("Failed to decrypt event after receiving room key: {e}");
+                            None
+                        }
+                    }
+                }
+                .instrument(info_span!(
+                    "retry_one",
+                    session_id = field::Empty,
+                    event_id = field::Empty
+                ))
+            };
 
-                let Some(remote_event) = event_item.as_remote() else {
-                    error!("Key for unable-to-decrypt timeline item is not an event ID");
-                    return None;
+            // Loop through all the indices, in order so we don't decrypt edits
+            // before the event being edited, if both were UTD. Keep track of
+            // index change as UTDs are removed instead of updated.
+            let mut offset = 0;
+            for idx in retry_indices {
```

Comment:
Before my last fixup commit, I was doing `.rev()` because it was easier to implement (no need for `offset`. However, that broke retrying when there were both a message and an edit to that message as UTDs. The original message has to be decrypted first for that to work.

Reply:
OK.